File: packages/core/src/growth/diagnose.ts
/**
 * Diagnose a Business Scorecard and route to the single highest-leverage constraint. Pure port of
 * Skills/growth-os/scripts/diagnose.py: work top-down (Market → Offer → Money Model → Leads), STOP
 * at the first failing gate, emit ONE prescription (CLAUDE.md §1, Convex-free).
 *
 * Deliberately conservative (§4 / RESEARCH Pitfall 4): when the money-model gate's financial inputs
 * are all null there is no basis to route OR to declare health, so it ASKS (naming the missing
 * figure) instead of fabricating a metric or a false "scale" verdict.
 */
import type { SpecialistRoute } from "../specialists";
import { FLOOR_RATIO, INDUSTRY_MULTIPLE } from "./financialSpine";
import type { Scorecard } from "./scorecard";

export type Prescription = {
  constraint: string;
  gate: 0 | 1 | 2 | 3 | "scale";
  /**
   * Closed to the dispatchable specialists (DISP-01). `""` stays representable ON PURPOSE — the
   * not-enough-data ask branch below emits it deliberately, and `resolveSpecialist("")` refuses it
   * at runtime rather than routing on a guess. Dependency direction is growth/ → specialists,
   * never the reverse.
   */
  route: SpecialistRoute | "";
  playbook: string;
  reason: string;
  proofMetric: string;
  roadmapLevel: number | null;
  /** Set => "not enough data, ask for this"; NEVER accompanied by a route or a fabricated metric. */
  ask?: string;
};

export function diagnose(sc: Scorecard): Prescription {
  const level = sc.position.roadmapLevel;
  const rx = (
    gate: Prescription["gate"],
    constraint: string,
    // Every rx() gate routes somewhere; only the ask branch below (which builds its object
    // literal directly) emits "". Narrower than Prescription["route"] on purpose.
    route: SpecialistRoute,
    playbook: string,
    reason: string,
    proofMetric: string,
  ): Prescription => ({ constraint, gate, route, playbook, reason, proofMetric, roadmapLevel: level });

  // --- Gate 0: Market (a great offer to the wrong crowd falls on deaf ears) ---
  if (sc.identity.marketViable === false) {
    return rx(
      0,
      "Market is not viable (no pain / no money / shrinking).",
      "offer-architect",
      "01-select-market",
      "A great offer to the wrong crowd falls on deaf ears. Fix the market/niche first.",
      "market passes the 4 indicators (pain, power, targetable, growing)",
    );
  }

  // --- Gate 1: Offer ---
  // An offer is "on file" when EITHER signal affirms one: the free-text list, or any
  // offerTypesPresent leaf the user affirmed (the checklist literally states "we sell an
  // attraction offer today"). Measured live (eval run 56bff5b8, fixture 31, 4/4 attempts):
  // the executive records the checklist booleans from prose but not the free-text list, and
  // reading the list alone diagnosed "no offer worth buying" against a scorecard affirming
  // two live offers — gate 1 fired before gate 2's offerTypeCount was ever consulted.
  // `=== true` mirrors the offerTypeCount checklist rule below (defence against non-boolean leaks).
  const hasOffer =
    sc.identity.currentOffers.length > 0 ||
    Object.values(sc.modelCard.offerTypesPresent).some((v) => v === true);
  const veScores = Object.values(sc.offerCard.valueEquation).filter(
    (v): v is number => typeof v === "number",
  );
  const weakOffer = veScores.length > 0 && Math.min(...veScores) <= 3;
  if (!hasOffer) {
    return rx(
      1,
      "No offer worth buying yet.",
      "offer-architect",
      "02-build-offer",
      "Nothing downstream matters without a differentiated offer.",
      "a stacked Grand Slam Offer exists with value >> price",
    );
  }
  if (sc.identity.commodity === true || weakOffer) {
    return rx(
      1,
      "Offer is a commodity / perceived value too low.",
      "offer-architect",
      "03-enhance-offer",
      "Buyers compare on price because value isn't differentiated. Raise value before spend.",
      "value-equation scores raised; conversion or price up",
    );
  }

  // --- Gate 2: Money Model ---
  const { thirtyDayPayback: payback } = sc.modelCard;
  const { cac, ltgp, thirtyDayCashPerCustomer: tdc, industryAvgCac: iac } = sc.financials;

  // Conservative: no financial signal at all → ask, never guess a route/metric or a false "scale".
  if (payback == null && tdc == null && cac == null && ltgp == null) {
    return {
      constraint: "Not enough financial data to diagnose the money model.",
      gate: 2,
      route: "",
      playbook: "",
      reason:
        "Vault-first, then ask: without CAC, LTGP, 30-day cash, or payback there is no basis to route or to declare the business healthy.",
      proofMetric: "",
      roadmapLevel: level,
      ask: "Provide 30-day cash per customer, CAC, LTGP, and whether acquisition cost is recovered within 30 days.",
    };
  }

  // `=== true`, not `filter(Boolean)`: these leaves are a CHECKLIST, and a non-boolean that slipped
  // past the write-site coercion (evaluations.ts) — notably the string "false", which is truthy —
  // would otherwise count a known-ABSENT offer type as PRESENT and silently skip this gate.
  // Defence in depth: unknown ⇒ the conservative "absent" default, never a wrong diagnosis.
  const offerTypeCount = Object.values(sc.modelCard.offerTypesPresent).filter(
    (v) => v === true,
  ).length;
  const ratio = ltgp && cac ? Math.round((ltgp / cac) * 100) / 100 : null;

  if (payback === false) {
    return rx(
      2,
      "30-day payback fails — cash is trapped.",
      "money-model-designer",
      "01-assess-money-model",
      "Acquisition cost isn't recovered in 30 days, so the business can't self-fund growth.",
      "30-day cash >= cost to get+service a customer (CFA achieved)",
    );
  }
  if (tdc != null && cac != null && tdc < cac) {
    return rx(
      2,
      "Customer doesn't pay for themselves in 30 days.",
      "money-model-designer",
      "06-assemble",
      "Pull cash forward with an upsell/downsell/continuity sequence.",
      "30-day cash >= CAC",
    );
  }
  if (offerTypeCount <= 1) {
    return rx(
      2,
      "Only one thing to sell — no upsell/downsell/continuity.",
      "money-model-designer",
      "06-assemble",
      "A sequence of offers multiplies 30-day cash from the same customers.",
      "attraction + upsell + downsell + continuity all present",
    );
  }
  if (ratio != null && ratio < FLOOR_RATIO && iac && cac && cac <= iac * INDUSTRY_MULTIPLE) {
    return rx(
      2,
      `LTGP:CAC ${ratio}:1 is below floor because LTGP is low.`,
      "money-model-designer",
      "01-assess-money-model",
      "CAC is within industry norms, so the ceiling is monetization — raise LTGP.",
      `LTGP:CAC >= ${FLOOR_RATIO}:1`,
    );
  }

  // --- Gate 3: Leads ---
  if (iac && cac && cac > iac * INDUSTRY_MULTIPLE) {
    return rx(
      3,
      `CAC (${cac}) exceeds 3x industry average (${iac}).`,
      "lead-engine",
      "11-more-better-new",
      "Advertising is inefficient — find the funnel constraint and test one thing per week.",
      "CAC falls below 3x industry average",
    );
  }
  // `=== true` for the same reason as offerTypeCount above — run 509373bf stored all four channels
  // as the STRING "false", counted 4 active, and made this gate unreachable.
  const activeChannels = Object.values(sc.leadCard.coreFourActive).filter((v) => v === true).length;
  if (activeChannels === 0) {
    return rx(
      3,
      "No lead channel is running.",
      "lead-engine",
      "01-pick-channel",
      "Offer and money model are sound; the constraint is that too few people know.",
      ">= 1 Core Four channel running at Rule-of-100 volume",
    );
  }

  // --- Nothing failed: scale ---
  return rx(
    "scale",
    "No failing gate — offer, money model, and leads are healthy.",
    "lead-engine",
    "11-more-better-new",
    "Compound the working channels (More/Better/New) and add Lead Getters.",
    "lead volume and LTGP:CAC both climbing",
  );
}

const gateOrder = (g: Prescription["gate"]): number => (g === "scale" ? 4 : g);

/** Order prescriptions by gate (Market<Offer<Money<Leads<Scale) — fix the ONE bottleneck first. */
export function leverageRank(scs: Prescription[]): Prescription[] {
  return [...scs].sort((a, b) => gateOrder(a.gate) - gateOrder(b.gate));
}

File: Skills/growth-os/scripts/diagnose.py
#!/usr/bin/env python3
"""Diagnose a Business Scorecard and route to the single highest-leverage skill.

Implements references/diagnostic-tree.md: work top-down (Market -> Offer -> Money Model
-> Leads), stop at the first failing gate, emit one prescription.

Usage:
  python diagnose.py --scorecard biz.json
  echo '{...scorecard json...}' | python diagnose.py --stdin

Output JSON: {constraint, gate, route, playbook, reason, proof_metric, roadmap_level}
The router is deliberately conservative: when a value is unknown it asks for it rather
than guessing, because a wrong diagnosis wastes the most time.
"""
import argparse
import json
import sys

FLOOR_RATIO = 3.0
INDUSTRY_MULTIPLE = 3.0


def _g(d, *path, default=None):
    cur = d
    for p in path:
        if not isinstance(cur, dict) or p not in cur or cur[p] is None:
            return default
        cur = cur[p]
    return cur


def diagnose(sc):
    missing = []

    # --- Gate 0: Market ---
    market_bad = _g(sc, "identity", "market_viable") is False
    if market_bad:
        return _rx("Market is not viable (no pain / no money / shrinking).", "market",
                   "offer-architect", "01-select-market",
                   "A great offer to the wrong crowd falls on deaf ears. Fix the market/niche first.",
                   "market passes the 4 indicators (pain, power, targetable, growing)",
                   _g(sc, "position", "roadmap_level"))

    # --- Gate 1: Offer ---
    has_offer = bool(_g(sc, "identity", "current_offers", default=[]))
    ve = _g(sc, "offer_card", "value_equation", default={}) or {}
    ve_scores = [v for v in ve.values() if isinstance(v, (int, float))]
    weak_offer = bool(ve_scores) and min(ve_scores) <= 3
    commodity = _g(sc, "identity", "commodity") is True
    if not has_offer:
        return _rx("No offer worth buying yet.", "offer", "offer-architect", "02-build-offer",
                   "Nothing downstream matters without a differentiated offer.",
                   "a stacked Grand Slam Offer exists with value >> price",
                   _g(sc, "position", "roadmap_level"))
    if commodity or weak_offer:
        pb = "03-enhance-offer" if has_offer else "02-build-offer"
        return _rx("Offer is a commodity / perceived value too low.", "offer", "offer-architect", pb,
                   "Buyers compare on price because value isn't differentiated. Raise value before spend.",
                   "value-equation scores raised; conversion or price up",
                   _g(sc, "position", "roadmap_level"))

    # --- Gate 2: Money Model ---
    payback = _g(sc, "model_card", "thirty_day_payback")
    tdc = _g(sc, "financials", "thirty_day_cash_per_customer")
    cac = _g(sc, "financials", "cac")
    ltgp = _g(sc, "financials", "ltgp")
    types = _g(sc, "model_card", "offer_types_present", default={}) or {}
    only_one_offer = sum(1 for v in types.values() if v) <= 1

    ratio = round(ltgp / cac, 2) if (ltgp and cac) else None

    if payback is False:
        return _rx("30-day payback fails — cash is trapped.", "money_model", "money-model-designer",
                   "01-assess-money-model",
                   "Acquisition cost isn't recovered in 30 days, so the business can't self-fund growth.",
                   "30-day cash >= cost to get+service a customer (CFA achieved)",
                   _g(sc, "position", "roadmap_level"))
    if tdc is not None and cac is not None and tdc < cac:
        return _rx("Customer doesn't pay for themselves in 30 days.", "money_model",
                   "money-model-designer", "06-assemble",
                   "Pull cash forward with an upsell/downsell/continuity sequence.",
                   "30-day cash >= CAC", _g(sc, "position", "roadmap_level"))
    if only_one_offer:
        return _rx("Only one thing to sell — no upsell/downsell/continuity.", "money_model",
                   "money-model-designer", "06-assemble",
                   "A sequence of offers multiplies 30-day cash from the same customers.",
                   "attraction + upsell + downsell + continuity all present",
                   _g(sc, "position", "roadmap_level"))
    if ratio is not None and ratio < FLOOR_RATIO:
        # low ratio: is it LTGP (model) or CAC (ads)?
        iac = _g(sc, "financials", "industry_avg_cac")
        if iac and cac and cac <= iac * INDUSTRY_MULTIPLE:
            return _rx(f"LTGP:CAC {ratio}:1 is below floor because LTGP is low.", "money_model",
                       "money-model-designer", "01-assess-money-model",
                       "CAC is within industry norms, so the ceiling is monetization — raise LTGP.",
                       f"LTGP:CAC >= {FLOOR_RATIO}:1", _g(sc, "position", "roadmap_level"))

    # --- Gate 3: Leads ---
    iac = _g(sc, "financials", "industry_avg_cac")
    if iac and cac and cac > iac * INDUSTRY_MULTIPLE:
        return _rx(f"CAC ({cac}) exceeds 3x industry average ({iac}).", "leads", "lead-engine",
                   "11-more-better-new",
                   "Advertising is inefficient — find the funnel constraint and test one thing per week.",
                   "CAC falls below 3x industry average", _g(sc, "position", "roadmap_level"))

    active_channels = sum(1 for v in (_g(sc, "lead_card", "core_four_active", default={}) or {}).values() if v)
    if active_channels == 0:
        return _rx("No lead channel is running.", "leads", "lead-engine", "01-pick-channel",
                   "Offer and money model are sound; the constraint is that too few people know.",
                   ">= 1 Core Four channel running at Rule-of-100 volume",
                   _g(sc, "position", "roadmap_level"))

    # --- Nothing failed: scale ---
    if not missing:
        return _rx("No failing gate — offer, money model, and leads are healthy.", "scale", "lead-engine",
                   "11-more-better-new",
                   "Compound the working channels (More/Better/New) and add Lead Getters.",
                   "lead volume and LTGP:CAC both climbing",
                   _g(sc, "position", "roadmap_level"))


def _rx(constraint, gate, route, playbook, reason, proof, level):
    return {
        "constraint": constraint,
        "gate": gate,
        "route": route,
        "playbook": playbook,
        "reason": reason,
        "proof_metric": proof,
        "roadmap_level": level,
    }


def main(argv=None):
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument("--scorecard")
    p.add_argument("--stdin", action="store_true")
    args = p.parse_args(argv)
    if args.stdin:
        sc = json.load(sys.stdin)
    elif args.scorecard:
        with open(args.scorecard, "r", encoding="utf-8") as f:
            sc = json.load(f)
    else:
        print(json.dumps({"error": "provide --scorecard PATH or --stdin"}))
        return 1
    print(json.dumps(diagnose(sc), indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
